File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    var head = &ListNode{}
    var tail = head
    for list1 != nil && list2 != nil {
        if list1.Val > list2.Val {
            tail.Next = list2
            list2 = list2.Next
        } else {
            tail.Next = list1
            list1 = list1.Next
        }
        tail = tail.Next
    }
    
    if list1 != nil {
        tail.Next = list1
    } else if list2 != nil {
        tail.Next = list2
    }
    
    return head.Next
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* curr1 = list1;
        ListNode* curr2 = list2;
        ListNode* ans = nullptr;

        while (curr1 && curr2) {
            int val;
            if (curr1->val < curr2->val) {
                val = curr1->val;
                curr1 = curr1->next;
            } else {
                val = curr2->val;
                curr2 = curr2->next;
            }

            ListNode* x = new ListNode(val);
            x->next = ans;
            ans = x;
        }

        while (curr1) {
            ListNode* x = new ListNode(curr1->val);
            x->next = ans;
            ans = x;
            curr1 = curr1->next;
        }

        while (curr2) {
            ListNode* x = new ListNode(curr2->val);
            x->next = ans;
            ans = x;
            curr2 = curr2->next;
        }

        ListNode* ans2 = nullptr;
        while (ans) {
            ListNode* x = new ListNode(ans->val);
            x->next = ans2;
            ans2 = x;
            ans = ans->next;
        }
        
        return ans2;
    }
};